src/frontend/src/components/dashboard/DashboardLayout.tsx

```diff
@@ -0,0 +1,290 @@
+import { Center, Divider, Loader } from '@mantine/core';
+import { useDisclosure, useHotkeys } from '@mantine/hooks';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { Layout, Responsive, WidthProvider } from 'react-grid-layout';
+
+import { useDashboardItems } from '../../hooks/UseDashboardItems';
+import { useUserState } from '../../states/UserState';
+import DashboardMenu from './DashboardMenu';
+import DashboardWidget, { DashboardWidgetProps } from './DashboardWidget';
+import DashboardWidgetDrawer from './DashboardWidgetDrawer';
+
+const ReactGridLayout = WidthProvider(Responsive);
+
+/**
+ * Save the dashboard layout to local storage
+ */
+function saveDashboardLayout(layouts: any, userId: number | undefined): void {
+  const data = JSON.stringify(layouts);
+
+  if (userId) {
+    localStorage?.setItem(`dashboard-layout-${userId}`, data);
```

Comment:
Wouldn't it make sense to store the dashboard layout on the server in a user setting?

Reply:
Great idea! I'll see how that would work

Reply:
I have this working, however the "layout" data does take up a lot of space. The `UserSetting.value` field currently has a restricted length of 2,000 characters. This equates to (roughly) 30 widgets - assuming a separate setting per "layout breakpoint" (i.e. "sm" / "md" / "lg" / "xl" which depend on the screen size).

This seems pretty limited IMO. We could bump up the length restriction on the "value" field, and also have a single setting for each layout "breakpoint".

Alternatively, we could live with the dashboard being cached to the browser (at least for now)?

Reply:
I'm going to leave it as is for this PR at least. Perhaps we can work on this as a future PR.